src/Microsoft.Health.Fhir.CosmosDb/Features/Search/FhirCosmosSearchService.cs

```diff
@@ -131,6 +162,128 @@ public class FhirCosmosSearchService : SearchService
+
+                if (!string.IsNullOrEmpty(chainedResults.continuationToken))
+                {
+                    throw new InvalidSearchOperationException(string.Format(Resources.ChainedExpressionSubqueryLimit, _chainedSearchMaxSubqueryItemLimit));
+                }
+
+                Expression[] chainedExpressionReferences;
+
+                if (!expression.Reversed)
+                {
+                    // When normal chained expression we can filter using references in the parent object. e.g. Observation.subject
+                    // The following expression constrains "subject" references on "Observation" with the ids that have matched the sub-query
+                    chainedExpressionReferences = chainedResults.results.Select(x =>
+                            Expression.SearchParameter(
+                                expression.ReferenceSearchParameter,
+                                Expression.And(
+                                    Expression.Equals(FieldName.ReferenceResourceId, null, x.Id),
+                                    Expression.Equals(FieldName.ReferenceResourceType, null, filteredType))))
+                        .ToArray<Expression>();
+                }
+                else
+                {
+                    // When reverse chained, we take the ids and types from the child object and use it to filter the parent objects.
+                    // e.g. Patient?_has:Group:member:_id=group1. In this case we would have run the query there Group.id = group1
+                    // and returned the indexed entries for Group.member. The following query will use these items to filter the parent Patient query.
+                    chainedExpressionReferences = chainedResults.results.SelectMany(x =>
+                            x.ReferencesToInclude.Select(include =>
+                                Expression.And(
+                                Expression.SearchParameter(
+                                    _resourceIdSearchParameter,
+                                    Expression.Equals(FieldName.TokenCode, null, include.ResourceId)),
+                                Expression.SearchParameter(
+                                    _resourceTypeSearchParameter,
+                                    Expression.Equals(FieldName.TokenCode, null, include.ResourceTypeName)))))
+                        .ToArray<Expression>();
+                }
+
+                return chainedExpressionReferences.Length > 1 ? Expression.Or(chainedExpressionReferences) : chainedExpressionReferences.FirstOrDefault();
+            }
+
+            // Loop over all chained expressions in the search query and pass them into the local recursion function
+            foreach (ChainedExpression item in chainedExpressions)
+            {
+                Expression recursiveLookup = await RecursiveLookup(item);
+
+                // When no results are returned from a chain we can break since all operands in the search should be met
```

Comment:
>we can break  [](start = 61, length = 13)

Can I have situation where ReverseLookup on chainedExpression would yeild (something, null, something)? In this case we would ignore second something

Reply:
So treat this example as we're all learning together :) So, given the case:

```
GET Patient?general-practitioner.name=Joe&general-practitioner.address-state=MN&general-practitioner.address-postcode=1234
```

https://www.hl7.org/fhir/search.html#chaining
> Note that chained parameters are applied independently to the target resource
> may return Patients cared for by Joe from CA and Jane from MN [and Smith from 1234]: no one practitioner need satisfy [all] conditions

---

So in the case above if the filtered results returned (Joe#123, null, Smith#789) we could not satisfy the second condition which means pseudo logic would be: 

```sql
SELECT Patient WHERE Patient.GeneralPractitioner IN (Joe#123)
AND Patient.GeneralPractitioner IN ()
AND Patient.GeneralPractitioner IN (Smith#789)
```

I think this then makes sense as soon as a condition cannot be met to short-circuit and return, thoughts / different interpretation?

Reply:
Thanks for great explanation.
If that's the case, why not actually do proper short-circuit and do `return null` instead of the `break` ?

---
In reply to: [585140401](https://github.com/microsoft/fhir-server/pull/1695#discussion_r585140401) [](ancestors = 585140401)